File: golang/core/medium/my_calendar_1/my_calendar_1.go
package main

import (
	"fmt"
	"sort"
)

type Interval struct {
	start int
	end   int
}

type MyCalendar struct {
	bookings []Interval
}

func Constructor() MyCalendar {
	return MyCalendar{
		bookings: []Interval{},
	}
}

func (mc *MyCalendar) Book(start int, end int) bool {
	newInterval := Interval{start: start, end: end}

	// binary search
	index := sort.Search(len(mc.bookings), func(i int) bool {
		return mc.bookings[i].start >= newInterval.start
	})

	// check overlap with prev booking
	if index > 0 && mc.bookings[index-1].end > newInterval.start {
		return false
	}

	// check overlap with next booking
	if index < len(mc.bookings) && mc.bookings[index].start < newInterval.end {
		return false
	}

	// insert the new interval
	// append a zero value to increase the slice size by one
	mc.bookings = append(mc.bookings, Interval{})
	// shift elements to make space for the new interval
	copy(mc.bookings[index+1:], mc.bookings[index:])
	// insert
	mc.bookings[index] = newInterval

	return true
}

func main() {
	myCalendar := Constructor()
	fmt.Println(myCalendar.Book(10, 20))
	fmt.Println(myCalendar.Book(15, 25))
	fmt.Println(myCalendar.Book(20, 30))
	fmt.Printf("%v\n", myCalendar)
}

File: typescript/data_structures/oop/my_calendar_1.ts
class MyCalendar {
  private bookings: number[][]
  constructor() {
    this.bookings = []
  }

  book(start: number, end: number): boolean {
    // binary search to find correct insertion index
    let left = 0
    let right = this.bookings.length
    while (left < right) {
      const mid = Math.floor((left + right) / 2)
      if (this.bookings[mid][0] < start) {
        left = mid + 1
      } else {
        right = mid
      }
    }

    // check overlap with prev booking
    if (left > 0 && this.bookings[left - 1][1] > start) {
      return false
    }
    // check overlap with next booking
    if (left < this.bookings.length && this.bookings[left][0] < end) {
      return false
    }

    this.bookings.splice(left, 0, [start, end])
    return true
  }
}

/**
* Your MyCalendar object will be instantiated and called as such:
* var obj = new MyCalendar()
* var param_1 = obj.book(start,end)
*/
const myCalendar = new MyCalendar()
console.log(myCalendar.book(10, 20))
console.log(myCalendar.book(15, 25))
console.log(myCalendar.book(20, 30))
console.log(myCalendar)